selectDynamic not encodes given name

## Compiler version
3.3.0

## Minimized code

```Scala
class X extends scala.reflect.Selectable:
  val `+` = "1"
  
 val x = X()
```

This one is not working:
```Scala
x.selectDynamic("+")
```

```Scala
Caused by: java.lang.NoSuchMethodException: Playground$X.+()
```

But this one works:

```Scala
x.selectDynamic("$plus")
```

## Expectation
This one works
```Scala
x.selectDynamic("+")
``` 


Also I wonder is there a way of encoding given name by hand? 
As I understood this blocks `scalamock` migration to Scala 3, since `MethodSymbol#name` returned encoded name, while `Symbol#name` gives raw name, so I can't use reflection to get a field, I need to encode it somehow.

--- a/compiler/src/dotty/tools/backend/sjs/JSCodeGen.scala
+++ b/compiler/src/dotty/tools/backend/sjs/JSCodeGen.scala
@@ -36,6 +36,7 @@ import dotty.tools.dotc.transform.sjs.JSSymUtils.*
 
 import JSEncoding.*
 import ScopedVar.withScopedVars
+import scala.reflect.NameTransformer
 
 /** Main codegen for Scala.js IR.
  *
@@ -4218,7 +4219,7 @@ class JSCodeGen()(using genCtx: Context) {
       }
     }
 
-    val methodName = MethodName.reflectiveProxy(methodNameStr, formalParamTypeRefs)
+    val methodName = MethodName.reflectiveProxy(NameTransformer.encode(methodNameStr), formalParamTypeRefs)
 
     js.Apply(js.ApplyFlags.empty, selectedValueTree, js.MethodIdent(methodName), actualArgs)(jstpe.AnyType)
   }

--- a/library/src/scala/reflect/Selectable.scala
+++ b/library/src/scala/reflect/Selectable.scala
@@ -21,7 +21,7 @@ trait Selectable extends scala.Selectable:
   final def selectDynamic(name: String): Any =
     val rcls = selectedValue.getClass
     try
-      val fld = rcls.getField(name).nn
+      val fld = rcls.getField(NameTransformer.encode(name)).nn
       ensureAccessible(fld)
       fld.get(selectedValue)
     catch case ex: NoSuchFieldException =>
@@ -35,7 +35,7 @@ trait Selectable extends scala.Selectable:
    */
   final def applyDynamic(name: String, paramTypes: Class[?]*)(args: Any*): Any =
     val rcls = selectedValue.getClass
-    val mth = rcls.getMethod(name, paramTypes*).nn
+    val mth = rcls.getMethod(NameTransformer.encode(name), paramTypes*).nn
     ensureAccessible(mth)
     mth.invoke(selectedValue, args.asInstanceOf[Seq[AnyRef]]*)
 

--- a/tests/run/i18612-a.scala
+++ b/tests/run/i18612-a.scala
@@ -1,0 +1,6 @@
+class X extends scala.reflect.Selectable:
+  val `+` = "1"
+
+@main def Test =
+  val x = X()
+  assert(x.selectDynamic("+") == "1")

--- a/tests/run/i18612-b.scala
+++ b/tests/run/i18612-b.scala
@@ -1,0 +1,6 @@
+class X extends scala.reflect.Selectable:
+  val plus = "1"
+
+@main def Test =
+  val x = X()
+  assert(x.selectDynamic("plus") == "1")

--- a/tests/run/i18612-c.scala
+++ b/tests/run/i18612-c.scala
@@ -1,0 +1,7 @@
+class X extends scala.reflect.Selectable:
+  def + = "1"
+
+@main def Test =
+  val x = X()
+  assert(x.selectDynamic("+") == "1")
+  assert(x.applyDynamic("+")() == "1")

--- a/tests/run/i18612-d.scala
+++ b/tests/run/i18612-d.scala
@@ -1,0 +1,7 @@
+class X extends scala.reflect.Selectable:
+  def plus = "1"
+
+@main def Test =
+  val x = X()
+  assert(x.selectDynamic("plus") == "1")
+  assert(x.applyDynamic("plus")() == "1")

